src/integrationTest/java/org/opensearch/security/DlsIntegrationTests.java

```diff
@@ -517,4 +597,110 @@ public void testAggregateAndComputeStarRatings() throws IOException {
+
+            assertThat(searchResponse, isSuccessfulSearchResponse());
+            assertThat(searchResponse, numberOfTotalHitsIsEqualTo(10));
+
+            // shows that roles are additive and the overlapping role with less filtering is used
+            assertThat(
+                searchResponse,
+                searchHitsContainDocumentsInAnyOrder(
+                    UNION_ROLE_TEST_DATA.keySet().stream().map(id -> Pair.of(UNION_TEST_INDEX_NAME, id)).collect(Collectors.toList())
+                )
+            );
+        }
+    }
+
+    @Test
+    public void testNonOverlappingRoleUnionSearchFiltering() throws Exception {
+        try (RestHighLevelClient restHighLevelClient = cluster.getRestHighLevelClient(TEST_ROLE_ONE_USER)) {
+            SearchRequest searchRequest = new SearchRequest(UNION_TEST_INDEX_NAME);
+            SearchResponse searchResponse = restHighLevelClient.search(searchRequest, DEFAULT);
+
+            assertThat(searchResponse, isSuccessfulSearchResponse());
+            assertThat(searchResponse, numberOfTotalHitsIsEqualTo(4));
+            assertThat(
+                searchResponse,
+                searchHitsContainDocumentsInAnyOrder(
+                    UNION_ROLE_TEST_DATA.entrySet()
+                        .stream()
+                        .filter(e -> e.getValue().get("sensitive").equals(false))
+                        .map(e -> Pair.of(UNION_TEST_INDEX_NAME, e.getKey()))
+                        .collect(Collectors.toList())
+                )
+            );
+        }
+
+        try (RestHighLevelClient restHighLevelClient = cluster.getRestHighLevelClient(TEST_ROLE_THREE_USER)) {
+            SearchRequest searchRequest = new SearchRequest(UNION_TEST_INDEX_NAME);
+            SearchResponse searchResponse = restHighLevelClient.search(searchRequest, DEFAULT);
+
+            assertThat(searchResponse, isSuccessfulSearchResponse());
+            assertThat(searchResponse, numberOfTotalHitsIsEqualTo(5));
+            assertThat(
+                searchResponse,
+                searchHitsContainDocumentsInAnyOrder(
+                    UNION_ROLE_TEST_DATA.entrySet()
+                        .stream()
+                        .filter(e -> e.getValue().get("genre").equals("History"))
+                        .map(e -> Pair.of(UNION_TEST_INDEX_NAME, e.getKey()))
+                        .collect(Collectors.toList())
+                )
+            );
+        }
+
+        try (RestHighLevelClient restHighLevelClient = cluster.getRestHighLevelClient(TEST_UNION_OF_NONOVERLAPPING_ROLES_USER)) {
+            SearchRequest searchRequest = new SearchRequest(UNION_TEST_INDEX_NAME);
+            SearchResponse searchResponse = restHighLevelClient.search(searchRequest, DEFAULT);
+
+            assertThat(searchResponse, isSuccessfulSearchResponse());
+            assertThat(searchResponse, numberOfTotalHitsIsEqualTo(9));
```

Comment:
This code can be cleaned up to allow re-usability.

```suggestion
            assertSearchResponseHitsEqualTo(9);
```

and a function:
```
private void assertSearchResponseHitsEqualTo(int hits) {
        SearchRequest searchRequest = new SearchRequest(UNION_TEST_INDEX_NAME);
        SearchResponse searchResponse = restHighLevelClient.search(searchRequest, DEFAULT);

        assertThat(searchResponse, isSuccessfulSearchResponse());
        assertThat(searchResponse, numberOfTotalHitsIsEqualTo(hits));
}
```

Reply:
searchResponse is used in the two checks below this so I'll keep them in the main function but I'll make that helper function to be used in future tests